File: models/Gpt-3.5-turbo/cpp/code/problems/range_queries/Pizzeria_Queries.cpp

#include <iostream>
#include <vector>
#include <climits>

using namespace std;

// Segment tree node structure
struct Node {
    int minPrice;
};

// Build segment tree
void build(vector<int>& prices, vector<Node>& tree, int node, int start, int end) {
    if (start == end) {
        tree[node].minPrice = prices[start];
        return;
    }

    int mid = (start + end) / 2;
    build(prices, tree, 2 * node, start, mid);
    build(prices, tree, 2 * node + 1, mid + 1, end);

    tree[node].minPrice = min(tree[2 * node].minPrice, tree[2 * node + 1].minPrice);
}

// Update pizza price in segment tree
void update(vector<Node>& tree, int node, int start, int end, int idx, int price) {
    if (start == end) {
        tree[node].minPrice = price;
        return;
    }

    int mid = (start + end) / 2;
    if (idx <= mid) {
        update(tree, 2 * node, start, mid, idx, price);
    } else {
        update(tree, 2 * node + 1, mid + 1, end, idx, price);
    }

    tree[node].minPrice = min(tree[2 * node].minPrice, tree[2 * node + 1].minPrice);
}

// Query minimum pizza price in segment tree
int query(vector<Node>& tree, int node, int start, int end, int left, int right) {
    if (start > right || end < left) {
        return INT_MAX;
    }

    if (left <= start && end <= right) {
        return tree[node].minPrice;
    }

    int mid = (start + end) / 2;
    int leftPrice = query(tree, 2 * node, start, mid, left, right);
    int rightPrice = query(tree, 2 * node + 1, mid + 1, end, left, right);

    return min(leftPrice, rightPrice);
}

int main() {
    int n, q;
    cin >> n >> q;

    vector<int> prices(n);
    for (int i = 0; i < n; ++i) {
        cin >> prices[i];
    }

    vector<Node> tree(4 * n);
    build(prices, tree, 1, 0, n - 1);

    while (q--) {
        int type, k, x;
        cin >> type >> k;
        
        if (type == 1) {
            cin >> x;
            update(tree, 1, 0, n - 1, k - 1, x);
        } else {
            cout << query(tree, 1, 0, n - 1, k - 1, k - 1) << endl;
        }
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/range_queries/Pizzeria_Queries.py

class Node:
    def __init__(self, start, end):
        self.start = start   # starting index of the range
        self.end = end       # ending index of the range
        self.min_price = float('inf')  # minimum pizza price within the range
        self.left = None     # left child node
        self.right = None    # right child node


def build_segment_tree(arr, start, end):
    if start == end:
        node = Node(start, end)
        node.min_price = arr[start]
        return node
    
    mid = (start + end) // 2
    left_child = build_segment_tree(arr, start, mid)
    right_child = build_segment_tree(arr, mid + 1, end)
    
    node = Node(start, end)
    node.left = left_child
    node.right = right_child
    node.min_price = min(left_child.min_price, right_child.min_price)
    
    return node


def update_segment_tree(node, index, new_price):
    if node.start == node.end:
        node.min_price = new_price
        return
    
    mid = (node.start + node.end) // 2
    if index <= mid:
        update_segment_tree(node.left, index, new_price)
    else:
        update_segment_tree(node.right, index, new_price)
        
    node.min_price = min(node.left.min_price, node.right.min_price)


def query_segment_tree(node, start, end):
    if start > node.end or end < node.start:
        return float('inf')
    
    if start <= node.start and end >= node.end:
        return node.min_price
    
    return min(query_segment_tree(node.left, start, end), query_segment_tree(node.right, start, end))


n, q = map(int, input().split())
prices = list(map(int, input().split()))

# Build the segment tree
root = build_segment_tree(prices, 0, n-1)

for _ in range(q):
    query = list(map(int, input().split()))
    if query[0] == 1:
        # Update pizza price
        update_segment_tree(root, query[1] - 1, query[2])
    else:
        # Query minimum pizza price within a range
        start, end = query[1] - 1, query[1] - 1
        min_price = query_segment_tree(root, start, end)
        print(min_price)
